herd/AArch64Sem.ml

```diff
@@ -1075,6 +1085,74 @@ module Make
             (fun loc v -> Act.tag_access quad Dir.W loc v) ii
         else m a
 
+(* Perform the authentication operation of the `AUT*` instruction, assume that
+ * the CPU can "predict" if the authentication will succede, such that in case
+ * of success, their is no basic-dep (only a pick-basic-dep) between the
+ * register-read event of the morifier, and the register-write event of the
+ * destination *)
+      let pauth2_with_fpac pointer modifier key ii mok mfault =
+        let mfault md mn =
+          (md >>| mn >>= fun _ -> commit_pred ii) >>*= fun _ -> mfault in
+
+        let original_pointer md =
+          md >>= M.op1 (Op.ArchOp1 AArch64Op.MakeCanonical) in
+
+        (* data_input_next allow to not have a control depedency between the
+         * commit event and the register-read event in Xd *)
+        let mok md mn =
+          (md >>| mn >>= fun _ -> commit_pred ii) >>*= fun _ ->
+          (M.data_input_next (original_pointer md) (fun ptr -> mok (M.unitT ptr)))
+        in
+
+        let auth_check xd xn =
+          M.op (Op.ArchOp (AArch64Op.AddPAC (false,key))) xd xn >>=
+          M.op1 (Op.ArchOp1 AArch64Op.CheckCanonical)
+        in
+
+        M.delay_kont "read pointer" pointer (fun xd md ->
+          M.delay_kont "read modifier" modifier (fun xn mn ->
+            auth_check xd xn >>= fun c ->
+            M.choiceT c
+              (mok md mn)
+              (mfault md mn)
+          )
+        )
+
+(* Perform the authentication operation of the `AUT*` instruction, assume that
+ * the CPU can "predict" if the authentication will succede, such that in case
+ * of success, their is no basic-dep (only a pick-basic-dep) between the
+ * register-read event of the morifier, and the register-write event of the
+ * destination *)
+      let pauth2_without_fpac pointer modifier key ii mop =
```

Comment:
This is very similar as `pauth2_with_fpac` with the only difference being the `mfault` vs `mfail`, could you fact the common code in a common function?